core/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenEndpointFilter.java

```diff
@@ -214,4 +222,57 @@ public Authentication convert(HttpServletRequest request) {
 					new OAuth2AuthorizationCodeAuthenticationToken(code, clientId, redirectUri);
 		}
 	}
+
+	/**
+	 * Validates that client is authenticated by {@link OAuth2ClientAuthenticationFilter}
+	 *
+	 * @see OAuth2ClientAuthenticationToken
+	 * @see OAuth2ClientAuthenticationFilter
+	 * @see <a target="_blank" href="https://tools.ietf.org/html/rfc6749#section-4.4">Section 4.4 Client Credentials Grant</a>
+	 */
+	private static class ClientCredentialsAuthenticationConverter implements Converter<HttpServletRequest, Authentication> {
+
+		private static final String SCOPE_PARAMETER_NAME = "scope";
+		private final OAuth2ScopeParser scopesParser = new OAuth2ScopeParser();
+
+		@Override
+		public Authentication convert(HttpServletRequest request) {
+			Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+			if (authentication == null) {
+				return null;
+			}
+
+			if (!OAuth2ClientAuthenticationToken.class.isAssignableFrom(authentication.getClass())) {
+				throwError(OAuth2ErrorCodes.INVALID_CLIENT, "client_id");
+			}
+
+			OAuth2ClientAuthenticationToken clientAuthenticationToken = (OAuth2ClientAuthenticationToken) authentication;
+			if (!clientAuthenticationToken.isAuthenticated()) {
+				throwError(OAuth2ErrorCodes.INVALID_CLIENT, "client_id");
+			}
+
+			Set<String> clientScopes = clientAuthenticationToken.getRegisteredClient().getScopes();
+
+			// "scope" is optional, see https://tools.ietf.org/html/rfc6749#section-4.4.2 and https://tools.ietf.org/html/rfc6749#section-3.3
+			// if "scope" is not provided, client scopes are used
+			String scopeParameter = request.getParameter(SCOPE_PARAMETER_NAME);
+			if (StringUtils.isEmpty(scopeParameter)) {
+				return new OAuth2ClientCredentialsAuthenticationToken(clientAuthenticationToken, clientScopes);
+			}
+
+			Set<String> requestedScopes = null;
+			try {
+				requestedScopes = scopesParser.parse(scopeParameter);
```

Comment:
Please move `scope` validation in `OAuth2ClientCredentialsAuthenticationProvider`

Reply:
Actually the more I thought about this one, I don't think we need to validate here. The `scope` validation should be done when the `RegisteredClient` is created - similar to `ClientRegistration` in `oauth2-client`. We know then that the configured scopes are valid and all we need to do is match the configured ones against the requested ones (as-is). So let's remove `OAuth2ScopeParser` and pass in the requested scopes as is, eg:

`new HashSet<>(Arrays.asList(StringUtils.delimitedListToStringArray(scope, " ")))`

Reply:
The validation itself (e.g. check that client has all requested scopes) is happening in the provider. The purpose of`OAuth2ScopeParser`  is only parsing `scope` string request param to a set according to RFC. It is slightly more complicated than `delimitedListToStringArray` so I decided to introduce more strict parser to cover special characters, double spaces, etc. I think it still has some value, unless you absolutely want to remove it. 

Maybe exception for cases when `scope` is not a valid string should be different then, just general BAD_REQUEST?

Reply:
I still feel that `scope` validation should happen when a client is registered, eg. when a `RegisteredClient` is created. This ensures that the `scope`(s) the client is allowed to request `RegisteredClient.scopes` are valid from the start. Even if the requested `scope` has invalid characters, it still won't match against the registered (valid) scope. So I feel this extra validation at runtime is redundant so please remove `OAuth2ScopeParser`. FYI, this validation already exists in `RegisteredClient.validateScopes()`